src/InsetOrderOptimizer.cpp

```diff
@@ -52,81 +54,54 @@ InsetOrderOptimizer::InsetOrderOptimizer(const FffGcodeWriter& gcode_writer,
-    int direction;
-    //If the entire wall is printed with the current extruder, print all of it.
-    if (wall_0_extruder_nr == wall_x_extruder_nr && wall_x_extruder_nr == extruder_nr)
-    {
-        //If printing the outer inset first, start with the lowest inset.
-        //Otherwise start with the highest inset and iterate backwards.
-        if(inset_direction == InsetDirection::OUTSIDE_IN)
-        {
-            start_inset = 0;
-            end_inset = paths.size();
-            direction = 1;
-        }
-        else //INSIDE_OUT or CENTER_LAST.
+    const bool use_one_extruder = wall_0_extruder_nr == wall_x_extruder_nr;
+    const bool current_extruder_is_wall_x = wall_x_extruder_nr == extruder_nr;
+
+    auto should_reverse = [&](){
+        if (use_one_extruder && current_extruder_is_wall_x)
         {
-            start_inset = paths.size() - 1;
-            end_inset = -1;
-            direction = -1;
+            // The entire wall is printed with the current extruder.
+            // Reversing the insets now depends on the inverse of the inset direction.
+            // If we want to print the outer insets first we start with the lowest and move forward
+            // otherwise we start with the highest and iterate back.
+            return ! outer_to_inner;
         }
-    }
-    //If the wall is partially printed with the current extruder, print the correct part.
-    else if (wall_0_extruder_nr != wall_x_extruder_nr)
+        // If the wall is partially printed with the current extruder we need to move forward
+        // for the outer wall extruder and iterate back for the inner wall extruder
+        return current_extruder_is_wall_x;
+    };  // Helper lambda to ensure that the reverse bool can be a const type
+    const bool reverse = should_reverse();
+
+    // Switches the begin()...end() forward iterator for a rbegin()...rend() reverse iterator
+    // I can't wait till we use the C++20 standard and have access to ranges and views
+    auto get_walls_to_be_added = [&](const bool reverse, const std::vector<VariableWidthLines>& paths)
     {
-        //If the wall_0 and wall_x extruders are different, then only include the insets that should be printed by the
-        //current extruder_nr.
-        if(extruder_nr == wall_0_extruder_nr)
+        if (paths.empty())
         {
-            start_inset = 0;
-            end_inset = 1; // Ignore inner walls
-            direction = 1;
+            return std::vector<const ExtrusionLine*>{};
         }
-        else if(extruder_nr == wall_x_extruder_nr)
+        if (reverse)
         {
-            start_inset = paths.size() - 1;
-            end_inset = 0; // Ignore outer wall
-            direction = -1;
+            if (use_one_extruder)
+            {
+                return wallsToBeAdded(paths.rbegin(), paths.rend()); // Complete wall with one extruder
+            }
+            return wallsToBeAdded(paths.rbegin(), std::prev(paths.rend())); // Ignore outer wall
```

Comment:
... good, except now the comments are switched around as well.

I'll change that too.